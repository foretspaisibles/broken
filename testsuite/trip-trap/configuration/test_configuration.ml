(* Test_configuration -- Test configuration parser

Author: Michael Grünewald
Date: Fri Aug  8 14:55:52 CEST 2014

Gasoline (https://github.com/michipili/gasoline)
This file is part of Gasoline

Copyright © 2014 Michael Grünewald

This file must be used under the terms of the CeCILL-B.
This source file is licensed as described in the file COPYING, which
you should have received as part of this distribution. The terms
are also available at
http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.txt *)
open Printf

(* A parser definition logging all events. *)
module Logger =
struct
  type t = unit

  let pos_to_lc pos =
    pos.Lexing.pos_lnum,
    pos.Lexing.pos_cnum - pos.Lexing.pos_bol

  let excerpt_to_string tag excerpt =
    let startpos = Configuration_parser.startpos excerpt in
    let endpos = Configuration_parser.endpos excerpt in
    let text = Configuration_parser.text excerpt in
    let startline, startcol =
      pos_to_lc startpos
    in
    let endline, endcol =
      pos_to_lc endpos
    in
    sprintf "%03d:%03d-%03d:%03d %s(%S)"
      startline startcol
      endline endcol
      tag text

  let comment _ exc =
    print_endline (excerpt_to_string "COMMENT" exc)

  let section _ path =
    let process exc =
      print_endline (excerpt_to_string "PATH_ELT" exc)
    in
    List.iter process path

  let binding _ key value =
    begin
      print_endline (excerpt_to_string "KEY" key);
      print_endline (excerpt_to_string "VALUE" value);
    end

  let parse_error _ errpos error =
    let errline, errcol =
      pos_to_lc errpos
    in
    printf "%03d:%02d PARSE_ERROR(%S)\n"
      errline errcol
      (Configuration_parser.error_to_string error)

end

module Configuration_logger =
  Configuration_parser.Make(Logger)


let () = Configuration_logger.parse_file () "example.conf"
