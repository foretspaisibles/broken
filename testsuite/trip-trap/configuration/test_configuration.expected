001:001-001:066 COMMENT("## This file demonstrate various features of configuration files.")
002:001-002:003 COMMENT("##")
003:001-003:070 COMMENT("## The various comments found in this file are rarely useful. You can")
004:001-004:034 COMMENT("## consider them as random noise.")
005:001-005:006 PATH_ELT("login")
006:003-006:011 KEY("greeting")
006:014-006:033 VALUE("Welcome to ${HOST}")
007:004-007:011 KEY("failure")
007:014-007:039 VALUE("##Failure##")
008:007-008:011 KEY("help")
008:014-010:044 VALUE("I cannot do much to help you, \n              but I can demonstrate how configuration values \n              can span over multiple lines!")
012:000-012:011 KEY("quoted-help")
012:013-012:159 VALUE("I still cannot do much to help you, but I can demonstrate how configuration values can span over multiple lines!")
014:001-014:002 PATH_ELT("*")
012:013-014:023 PATH_ELT("User defined label")
015:001-015:018 COMMENT(" Another comment.")
016:000-016:018 KEY("ask-before-delete?")
016:021-016:027 VALUE("maybe")
018:001-018:005 PATH_ELT("file")
018:006-018:020 PATH_ELT("classification")
019:000-019:005 KEY("*.cpp")
019:008-019:021 VALUE("Useless junk")
020:000-020:005 KEY("*.hpp")
020:008-020:021 VALUE("Useless junk")
021:002-021:005 KEY("*.c")
021:008-021:026 VALUE("Good old C source")
022:002-022:005 KEY("*.h")
022:008-022:026 VALUE("Good old C header")
023:001-023:005 KEY("*.ml")
023:008-023:026 VALUE("OCaml source file")
024:000-024:005 KEY("*.mli")
024:008-024:029 VALUE("OCaml interface file")
025:000-025:005 KEY("*.mll")
025:008-025:020 VALUE("OCaml lexer")
026:000-026:005 KEY("*.mly")
026:008-026:021 VALUE("OCaml parser")
